h5pt.go

```diff
@@ -81,40 +93,172 @@ func (t *Table) ReadPackets(start, nrecords int, data interface{}) error {
+		// by another 64 bytes.
+		if memSize < (dataSize + offset) {
+			memSize += 64
+			ptr = C.realloc(ptr, memSize)
+			C.memset(unsafe.Pointer(uintptr(ptr)+uintptr(offset)), 0, 64)
+		}
+		C.memcpy(unsafe.Pointer(uintptr(ptr)+uintptr(offset)), dataPtr, dataSize)
+	}
+
+	return ptr, nil
+}
+
 // Append appends packets to the end of a packet table.
-func (t *Table) Append(data interface{}) error {
+func (t *Table) Append(data interface{}) (err error) {
 	rv := reflect.Indirect(reflect.ValueOf(data))
+	rp := reflect.Indirect(reflect.ValueOf(&data))
 	rt := rv.Type()
-	c_nrecords := C.size_t(0)
-	c_data := unsafe.Pointer(nil)
+	cNrecords := C.size_t(1)
+	cData := unsafe.Pointer(nil)
 
 	switch rt.Kind() {
+	case reflect.Slice, reflect.Array:
+		for i := 0; i < rv.Len(); i++ {
+			if err = t.Append(rv.Index(i).Interface()); err != nil {
+				return err
+			}
+		}
+		return nil
 
-	case reflect.Array:
-		c_nrecords = C.size_t(rv.Len())
-		c_data = unsafe.Pointer(rv.UnsafeAddr())
-
-	case reflect.Slice:
-		c_nrecords = C.size_t(rv.Len())
-		slice := (*reflect.SliceHeader)(unsafe.Pointer(rv.UnsafeAddr()))
-		c_data = unsafe.Pointer(slice.Data)
+	case reflect.Struct:
+		if cData, err = extractStructValues(rv, rt); err != nil {
+			return err
+		}
+		t.structs = append(t.structs, cData)
 
 	case reflect.String:
-		c_nrecords = C.size_t(rv.Len())
-		str := (*reflect.StringHeader)(unsafe.Pointer(rv.UnsafeAddr()))
-		c_data = unsafe.Pointer(str.Data)
+		stringData := C.CString(rv.String())
+		cData = unsafe.Pointer(&stringData)
+		t.strings = append(t.strings, (*C.char)(unsafe.Pointer(stringData)))
 
 	case reflect.Ptr:
-		c_nrecords = C.size_t(1)
-		c_data = unsafe.Pointer(rv.Elem().UnsafeAddr())
+		ptrVal := rp.Elem()
+		cData = unsafe.Pointer(&ptrVal)
+
+	case reflect.Bool:
+		val := C.uchar(0)
+		if data.(bool) {
+			val = 1
+		}
+		cData = unsafe.Pointer(&val)
+
+	case reflect.Int8:
+		val := C.int8_t(data.(int8))
+		cData = unsafe.Pointer(&val)
+
+	case reflect.Uint8:
+		val := C.uint8_t(data.(uint8))
+		cData = unsafe.Pointer(&val)
+
+	case reflect.Uint16:
+		val := C.uint16_t(data.(uint16))
+		cData = unsafe.Pointer(&val)
+
+	case reflect.Int16:
+		val := C.int16_t(data.(int16))
+		cData = unsafe.Pointer(&val)
+
+	case reflect.Int32:
+		val := C.int32_t(data.(int32))
+		cData = unsafe.Pointer(&val)
+
+	case reflect.Uint32:
+		val := C.uint32_t(data.(uint32))
+		cData = unsafe.Pointer(&val)
+
+	case reflect.Int64:
+		val := C.int64_t(data.(int64))
+		cData = unsafe.Pointer(&val)
+
+	case reflect.Uint64:
+		val := C.uint64_t(data.(uint64))
+		cData = unsafe.Pointer(&val)
+
+	case reflect.Float32:
+		val := C.float(data.(float32))
+		cData = unsafe.Pointer(&val)
+
+	case reflect.Float64:
+		val := C.double(data.(float64))
+		cData = unsafe.Pointer(&val)
 
 	default:
-		c_nrecords = C.size_t(1)
-		c_data = unsafe.Pointer(rv.UnsafeAddr())
+		err = fmt.Errorf("hdf5: PT Append does not support datatype (%s).", rt.Kind())
```

Comment:
if we take this branch of the `switch`, it seems to me like we drop this error on the floor as it's overriden later on, at line 259.

Reply:
Yes. You are right. Will need to update the returns.